### ISSUE TITLE
[Feature]: Model confidence calibration — beyond accuracy to reliability

### ISSUE THREAD

**Issue by Adit-Jain-srm** [Adit-Jain-srm/NightmareNet#28]
## Problem

NightmareNet's robustness score measures whether the model gives the **correct** answer under perturbation. But it doesn't measure whether the model **knows when it doesn't know**. A model that confidently outputs the wrong answer under adversarial attack is worse than one that says 'I'm not sure' - because downstream systems trust high-confidence outputs.

Current state:
- Evaluation outputs: accuracy, robustness_score (correct under distortion)
- Missing: calibration metrics (is confidence correlated with correctness?)
- Missing: uncertainty detection (does the model flag unreliable predictions?)
- Missing: the 'I don't know' option (abstention under uncertainty)

### 1. Expected Calibration Error (ECE) Metric

```python
# nightmarenet/evaluation/calibration.py
def compute_ece(predictions: List[float], labels: List[int], n_bins: int = 15) -> float:
    """Expected Calibration Error: measures gap between predicted confidence
    and actual accuracy across binned confidence levels.
    ECE = 0: perfectly calibrated (90% confidence -> 90% correct)
    ECE > 0.1: poorly calibrated (needs post-hoc correction)
    """
```

### 2. Temperature Scaling Post-Hoc Calibration

```python
# nightmarenet/evaluation/temperature_scaling.py
class TemperatureScaler:
    def fit(self, logits: Tensor, labels: Tensor) -> float:
        """Learn optimal temperature T on validation set.
        T > 1: softens overconfident predictions
        T < 1: sharpens underconfident predictions
        """
    
    def calibrate(self, logits: Tensor) -> Tensor:
        """Apply learned temperature to new predictions."""
```

### 3. Uncertainty-Aware Evaluation

Extend the robustness evaluation to track confidence degradation:

```python
class EvaluationResult:
    accuracy: float
    robustness_score: float
    ece_before_calibration: float
    ece_after_calibration: float
    abstention_rate: float  # % of inputs where confidence < threshold
    confidence_degradation: float  # avg confidence drop under distortion
```

### 4. Abstention (Selective Prediction)

When model confidence drops below a configurable threshold after distortion:
- Flag the input as 'uncertain' rather than forcing a prediction
- Report abstention rate at each distortion strength
- Trade-off: higher threshold = more abstentions but fewer wrong answers

```yaml
evaluation:
  calibration:
    enabled: true
    temperature_scaling: true
    ece_bins: 15
  abstention:
    enabled: true
    confidence_threshold: 0.6
    report_coverage_accuracy_curve: true
```

### 5. Calibration Plot in Frontend

Add to Robustness Radar (Panel 06) or as a new sub-panel:
- Reliability diagram: predicted confidence (x) vs actual accuracy (y)
- Perfect calibration = diagonal line
- Show before/after temperature scaling
- Coverage-accuracy curve

### 6. Module Structure

```
nightmarenet/
  evaluation/
    calibration.py          # ECE computation, reliability diagram data
    temperature_scaling.py  # Post-hoc calibration (fit + transform)
    selective_prediction.py # Abstention logic, coverage-accuracy curves
```

## Technical Considerations

- Temperature scaling requires a held-out calibration set (not the same as test set)
- ECE is computed AFTER softmax (on probability space, not logit space)
- For generative models (GPT-2), calibration applies to per-token confidence
- Temperature T is a single scalar learned via NLL minimization (fast: < 1 second)
- Abstention threshold should be configurable per deployment context

## Research Context

- Guo et al. 2017 'On Calibration of Modern Neural Networks' - deep nets are poorly calibrated
- Stutz et al. 2020 'Confidence-Calibrated Adversarial Training' - calibration degrades under attack
- NightmareNet can show cyclic training produces BETTER calibrated models (testable hypothesis)

## Acceptance Criteria

- [ ] ECE metric computed during evaluation (before and after calibration)
- [ ] Temperature scaling fitted on validation split
- [ ] Abstention rate reported at each distortion strength
- [ ] Coverage-accuracy curve data generated (JSON)
- [ ] CLI `nightmarenet evaluate` includes calibration metrics in output
- [ ] YAML config for calibration settings
- [ ] Calibration works for classification models (softmax outputs)
- [ ] Reliability diagram data exposed for frontend
- [ ] Tests: ECE correctness, temperature scaling convergence, abstention logic

**Comment by Samm-05** [comment]
Hi @Adit-Jain-srm,

I'd like to work on this calibration and uncertainty evaluation feature. I have gone through the proposal and would like to implement it in phases (ECE → Temperature Scaling → Abstention → CLI & Tests) to keep the PRs reviewable.

Could you please assign this issue to me?

**Comment by Adit-Jain-srm** [comment]
Assigned, go ahead @Samm-05

**Comment by Adit-Jain-srm** [comment]
@Samm-05  It has been 4 days without activity on issue or opening PR. Please update your progress to prevent un-assignment of issue.

**Comment by Samm-05** [comment]
Hi @Adit-Jain-srm,

Apologies for the lack of updates over the past few days. Due to some unforeseen commitments, I wasn't able to make the progress I had planned on this issue.

I'm currently prioritizing and wrapping up my previously assigned tasks across projects. Once those are completed, I'll immediately resume work on this issue and provide regular progress updates. I remain interested in implementing this feature and appreciate your patience.

If you need to reassign the issue due to project timelines, I completely understand. Thank you for your understanding.

**Comment by Adit-Jain-srm** [comment]
Closing: no progress in 10 days. Replaced by a new issue with phased scope and updated format.
